bgcval2/functions/circulation.py

```diff
@@ -0,0 +1,389 @@
+loadedArea     = False
+loadedAltMask     = False
+
+
+def loadDataMask(gridfn,maskname,):
+    global umask_drake
+    global e2u_drake
+    global loadedArea
+    global e3v_AMOC26N
+    global e1v_AMOC26N
+    global tmask_AMOC26N    
+    if isinstance(gridfn, list) and len(gridfn)==1:
+        gridfn = gridfn[0]
+
+    nc = dataset(gridfn,'r')        
+    e2u_drake = nc.variables['e2u'][LAT0:LAT1,LON]
+    umask_drake = nc.variables['umask'][:,LAT0:LAT1,LON]
+    e3v_AMOC26N = nc.variables['e3v'][:,latslice26Nnm,:]   # z level height 3D
+    e1v_AMOC26N = nc.variables['e1v'][latslice26Nnm,:]     #
+    tmask_AMOC26N = nc.variables['tmask'][:,latslice26Nnm,:]    
+    nc.close()
+    loadedArea = True
+
+
+def loadAtlanticMask(altmaskfile,maskname='tmaskatl',):
+    global alttmask_AMOC26N
+    global loadedAltMask
+    nc = dataset(altmaskfile,'r')        
+    alttmask_AMOC26N = nc.variables[maskname][latslice26Nnm,:]   # 2D Atlantic mask
+ #       alttmask_AMOC26N[228,180:208]=0.
+    nc.close()
+    loadedAltMask = True
+
+
+def drakePassage(nc,keys,**kwargs):
+    """
+    This function. 
+    
+    nc: a netcdf openned as a dataset.
+    keys: a list of keys to use in this function.
+    
+    """
+    
+    if 'areafile' not in list(kwargs.keys()):
+        raise AssertionError("drakePassage:\t Needs an `areafile` kwarg to run calculation.")    
+
+    try:    maskname = kwargs['maskname']
+    except:    maskname = 'tmask'
+            
+    if not loadedArea: loadDataMask(kwargs['areafile'],maskname)
+    try:    e3u = nc.variables['thkcello'][0,:,LAT0:LAT1,LON]    
+    except:    e3u = nc.variables['e3u'     ][0,:,LAT0:LAT1,LON]
+
+    try:    velo = nc.variables['uo' ][0,:,LAT0:LAT1,LON]    
+    except:    velo = nc.variables['u3d'][0,:,LAT0:LAT1,LON]
+
+    drake = np.sum(velo*e3u*e2u_drake*umask_drake)*1.e-6
+    
+    return drake
+
+
+
+drakedetails     = {}
+def loadDrakeDetails(gridfn):
+    global drakedetails
+    print("cmip5DrakePassage:\topening grid file", gridfn)    
+    nc = dataset(gridfn,'r')
+    if 'Drake_A' in list(nc.variables.keys()): 
+        drakeKey = 'Drake_A'
+    elif 'Drake' in list(nc.variables.keys()): 
+        drakeKey = 'Drake'
+            
+    Drake_A = nc.variables[drakeKey][:]        
+    tmask     = nc.variables['tmask'][:]
+    drakedetails[(gridfn,'Drake_A')] = Drake_A
+    drakedetails[(gridfn,'tmask')]     = tmask
+    nc.close()
+       
+ 
+def cmip5DrakePassage(nc,keys,**kwargs):
+    if 'gridfile' not in list(kwargs.keys()):
+        raise AssertionError("drakePassage:\t Needs an `gridFile` kwarg to run calculation.")    
+    gridFile_u =     kwargs['gridfile']
+    try:    
+        drakexsectArea     = drakedetails[(gridFile_u,'Drake_A')]
+        tmasku        = drakedetails[(gridFile_u,'tmask')]
+    except:
+        loadDrakeDetails(gridFile_u)
+        drakexsectArea     = drakedetails[(gridFile_u,'Drake_A')]
+        tmasku        = drakedetails[(gridFile_u,'tmask')]
```

Comment:
the `try` might fail for one or two causes, in which case `except` kicks in and does a whole lot of things, that might fail too - think of the Traceback one might get from that. It's really high time you unpacked these cases one by one and attached an error raising it whenever it's needed, this is a bomb to debug :grin: